### PR TITLE
(CDAP-13418) Spark runtime in cloud

### PR DESCRIPTION
**Pull request by chtyim**
The changes has been broken down into commits. Please review one by one.

1. Handle exception from user spark program correctly to allow multiple attempts
   - This is a bug fix discovered along the way
2. Don’t store key content in memory and don’t serialize them
  - Yet another bug fix
3. Remove instance dependency on TokenSecureStoreRenewer
   - Just simplifying dependencies
4. Support Spark in Cloud
  - This is the final integration work

### CODE REVIEW

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-DUT6436-7

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkClassRewriter.java [hunk 1]
```diff
@@ -597,57 +604,84 @@ public MethodVisitor visitMethod(int access, String name, String desc, String si
         }
 
         // Wrap the original main method with
-        // Cancellable cancel = SparkRuntimeUtils.initSparkMain();
+        // CompletableFuture<?> completion = SparkRuntimeUtils.initSparkMain();
```
Comment (albertshau) [inline cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkClassRewriter.java]: this and other comments are inconsistent with the actual API

File: cdap-app-fabric/src/main/resources/setupSpark.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+# A simple python that runs with spark-submit to grab all SPARK environment variables
+import os
+for key, val in os.environ.items():
+    if key.startswith("SPARK_") or key.startswith("HADOOP_"):
+        print("export {}=\"{}\"".format(key, val))
```
Comment (albertshau) [inline cdap-app-fabric/src/main/resources/setupSpark.py]: missing newline at end of file

File: cdap-app-fabric/src/main/resources/setupSpark.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+#
+# Shell script to optionally setup spark environment variables for CDAP to use
+#!/bin/bash
+
+if [[ $(which spark-submit 2>/dev/null) ]]; then
+  spark-submit --master local[2] $1
+fi
```
Comment (albertshau) [inline cdap-app-fabric/src/main/resources/setupSpark.sh]: missing newline at end of file

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextProvider.java [hunk 4]
```diff
@@ -167,49 +169,56 @@ private static synchronized SparkRuntimeContext createIfNotExists() {
...
-                                                                                    metricsCollectionService,
-                                                                                    kafkaClientService,
-                                                                                    zkClientService);
-          try {
-            List<ListenableFuture<Service.State>> futures = future.get(5, TimeUnit.SECONDS);
-            System.out.println("SparkClassLoader services shutdown completed: " + futures);
-          } catch (Exception e) {
-            System.err.println("Exception when shutting down services");
-            e.printStackTrace(System.err);
+          System.out.println("Shutting Spark runtime services");
```
Comment (albertshau) [inline cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextProvider.java]: Shutting down

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkClassRewriter.java (hunk 1 above)
Comment (chtyim) [reply cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkClassRewriter.java]: fixed

File: cdap-app-fabric/src/main/resources/setupSpark.py (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/resources/setupSpark.py]: fixed

File: cdap-app-fabric/src/main/resources/setupSpark.sh (hunk 3 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/resources/setupSpark.sh]: fixed

File: cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextProvider.java (hunk 4 above)
Comment (chtyim) [reply cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextProvider.java]: fixed

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
Squashed commits into logical units. Merging.

**Pull request merged**